# Limpieza del dataset de bienes raíces

Este es un conjunto de datos (dataset) reales que fue descargado usando técnicas de web scraping. El archivo contiene registros de **Fotocasa**, el cual es uno de los sitios más populares de bienes raíces en España. Contiene miles de datos de casas reales publicadas en la web www.fotocasa.com.

El dataset fue descargado hace algunos años y en ningún caso se obtuvo beneficio económico de ello.

Tu objetivo es extraer tanta información como sea posible con el conocimiento que tienes hasta ahora de ciencia de datos.

In [1]:
import pandas as pd

#### Ejercicio 00

Lee el dataset data/real_estate.csv e intenta visualizar la tabla (★☆☆)

In [2]:
# Este archivo CSV contiene puntos y comas en lugar de comas como separadores
df = pd.read_csv('../data/real_estate.csv', sep=';')
df

Unnamed: 0  id_realEstates  isNew                    realEstate_name  \
0               1       153771986  False            ferrari 57 inmobiliaria   
1               2       153867863  False  tecnocasa fuenlabrada ferrocarril   
2               3       153430440  False                 look find boadilla   
3               4       152776331  False  tecnocasa fuenlabrada ferrocarril   
4               5       153180188  False            ferrari 57 inmobiliaria   
...           ...             ...    ...                                ...   
15330       15331       153901377  False                infocasa consulting   
15331       15332       150394373  False                inmobiliaria pulpon   
15332       15333       153901397  False             tecnocasa torrelodones   
15333       15334       152607440  False                inmobiliaria pulpon   
15334       15335       153901356  False                infocasa consulting   

       phone_realEstate                                       url_inmueble  \
0           912177526.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
1           916358736.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
2           916350408.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
3           916358736.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
4           912177526.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
...                 ...                                                ...   
15330       911360461.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15331       912788039.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15332       912780348.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15333       912788039.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
15334       911360461.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   

       rooms  bathrooms  surface   price  ... level4Id level5Id level6Id  \
0        3.0        2.0    103.0  195000  ...        0        0        0   
1        3.0        1.0      NaN   89000  ...        0        0        0   
2        2.0        2.0     99.0  390000  ...        0        0        0   
3        3.0        1.0     86.0   89000  ...        0        0        0   
4        2.0        2.0    106.0  172000  ...        0        0        0   
...      ...        ...      ...     ...  ...      ...      ...      ...   
15330    2.0        1.0     96.0  259470  ...        0        0        0   
15331    3.0        1.0    150.0  165000  ...        0        0        0   
15332    4.0        2.0    175.0  495000  ...        0        0        0   
15333    3.0        2.0    101.0  195000  ...        0        0        0   
15334    3.0        2.0    152.0  765000  ...        0        0        0   

      level7Id level8Id accuracy          latitude          longitude zipCode  \
0            0        0        0  40,2948276786438  -3,44402412135624     NaN   
1            0        0        1          40,28674           -3,79351     NaN   
2            0        0        0  40,4115646786438  -3,90662252135624     NaN   
3            0        0        0  40,2853785786438  -3,79508142135624     NaN   
4            0        0        0  40,2998774864376  -3,45226301356237     NaN   
...        ...      ...      ...               ...                ...     ...   
15330        0        0        0          40,45416           -3,70286     NaN   
15331        0        0        0          40,36652           -3,48951     NaN   
15332        0        0        0          40,57444           -3,92124     NaN   
15333        0        0        0          40,36967           -3,48105     NaN   
15334        0        0        0          40,45773           -3,69068     NaN   

      customZone  
0            NaN  
1            NaN  
2            NaN  
3            NaN  
4            NaN  
...          ...  
15330        NaN  
15331        NaN  
15332        NaN  
15333        NaN  
15334        NaN  

[15335

## Trabajando con un DataFrame

#### Ejercicio 01

¿Cuál es la casa más cara del dataset? (★☆☆)

Imprime la dirección y el precio de la casa seleccionada. Para visualizar el resultado utiliza un f-string. Por ejemplo:

```py
f'La casa más cara se encuentra en la dirección: {address} y su precio es {price} €'
```

In [3]:
max_price = df['price'].max()

casa_mas_cara = df[df['price'] == max_price]

address = casa_mas_cara.iloc[0]['address']
price = casa_mas_cara.iloc[0]['price']

f'La casa más cara está en la dirección: {address} y su precio es {price} €'

'La casa más cara está en la dirección: El Escorial y su precio es 8500000 €'

#### Ejercicio 02

¿Cuál es la casa más barata del dataset? (★☆☆)

Imprime la dirección y el precio de la casa seleccionada utilizando f-string

In [8]:
min_price = df['price'].min()

casa_mas_barata = df[df['price'] == min_price]

address = casa_mas_barata.iloc[0]['address']
price = casa_mas_barata.iloc[0]['price']

f'La casa más barata está en la dirección: {address} y su precio es {price} €'

La casa más barata está en la dirección: Parla y su precio es 0 €


Como comentario, veo en el dataset que hay varias casas que el precio es 0 €. Agregué solo una con iloc porque vi que es lo que solicitaba el ejercicio

#### Ejercicio 03

¿Cuál es la casa más grande del dataset? (★☆☆)

Imprime la dirección y el área de las casas seleccionadas utilizando f-string

In [11]:
max_surface = df['surface'].max()

casa_mas_grande = df[df['surface'] == max_surface]

address = casa_mas_grande.iloc[0]['address']
surface = casa_mas_grande.iloc[0]['surface']

f'La casa más grande está en la dirección: {address} y su area es {surface}'

La casa más grande está en la dirección: Sevilla la Nueva y su area es 249000.0


#### Ejercicio 04

¿Cuál es la casa más pequeña del dataset? (★☆☆)

In [12]:
min_surface = df['surface'].min()

casa_mas_chica = df[df['surface'] == min_surface]

address = casa_mas_chica.iloc[0]['address']
surface = casa_mas_chica.iloc[0]['surface']

f'La casa más chica está en la dirección: {address} y su area es {surface}'

La casa más chica está en la dirección: Calle Amparo,  Madrid Capital y su area es 15.0


#### Ejercicio 05.

¿El dataset contiene valores no admitidos (NAs)? (★☆☆)

- Muestra el nombre de las filas seguidas por un booleano (`True` o `False`) según contengan o no contengan NAs.
- También muestra el nombre de las columnas seguidas por un booleano (`True` o `False`) según contengan o no contengan NAs.

In [8]:
f"¿La fila tiene un valor NA?"
df.isna().any(axis=1)

0        True
1        True
2        True
3        True
4        True
         ... 
15330    True
15331    True
15332    True
15333    True
15334    True
Length: 15335, dtype: bool

In [12]:
# 3️⃣ Ver si hay valores nulos (NAs) en cada columna
f"¿Hay valores NA en las columnas?"
df.isna().any()

Unnamed: 0          False
id_realEstates      False
isNew               False
realEstate_name      True
phone_realEstate     True
url_inmueble        False
rooms                True
bathrooms            True
surface              True
price               False
date                False
description          True
address             False
country             False
level1              False
level2              False
level3              False
level4               True
level5              False
level6               True
level7               True
level8               True
upperLevel          False
countryId           False
level1Id            False
level2Id            False
level3Id            False
level4Id            False
level5Id            False
level6Id            False
level7Id            False
level8Id            False
accuracy            False
latitude            False
longitude           False
zipCode              True
customZone           True
dtype: bool

Todas las filas tienen por lo menos un valor NA

#### Ejercicio 06.

Elimina los NAs del dataset, si aplica (★★☆)

Muestra las dimensiones del DataFrame original y del DataFrame después de las eliminaciones.

In [10]:
f"Dimensiones originales: {df.shape}"

'Dimensiones originales: (15335, 37)'

In [11]:
df_sin_columnas = df.drop(['zipCode', 'customZone'], axis=1)

f"Dimensiones después de eliminar zipCode y customZone: {df_sin_columnas.shape}"

'Dimensiones después de eliminar zipCode y customZone: (15335, 35)'

Dejo el codigo planteado. Para mi no tiene sentido eliminar todos los NA del dataset (por filas con valor NA) porque nos quedariamos sin datos. Y tampoco eliminar las columnas con NA porque eliminariamos columnas que para cierta fila si tienen información. Por lo que veo las columnas zipCode y customZone son las que tienen el 100% de sus datos NA así que son las que eliminaria para mejorar el DataFrame

#### Ejercicio 07

¿Cuantas poblaciones (columna level5) contiene el dataset? (★☆☆)

- Muestra una lista con los nombres de las poblaciones
- Muestra el total de las mismas

In [15]:
poblaciones = df['level5'].unique()

len(poblaciones), poblaciones

(168,
 array(['Arganda del Rey', 'Fuenlabrada', 'Boadilla del Monte',
        'Las Rozas de Madrid', ' Madrid Capital', 'Villaviciosa de Odón',
        'Pinto', 'Valdemoro', 'Navalcarnero', 'Pozuelo de Alarcón',
        'Torrejón de Ardoz', 'Navalagamella', 'San Sebastián de los Reyes',
        'Rivas-vaciamadrid', 'Alpedrete', 'Móstoles',
        'San Fernando de Henares', 'Coslada', 'Becerril de la Sierra',
        'Alcalá de Henares', 'Chinchón', 'Parla', 'Alcorcón',
        'El Escorial', 'Leganés', 'Pedrezuela', 'Majadahonda',
        'Villanueva de la Cañada', 'Villanueva del Pardillo',
        'Torrelodones', 'Moralzarzal', 'Mejorada del Campo', 'Aranjuez',
        'Corpa', 'Getafe', 'Velilla de San Antonio', 'Sevilla la Nueva',
        'San Martín de la Vega', 'Villalbilla', 'Collado Villalba',
        'Alcobendas', 'El Molar (Madrid)', 'Moraleja de Enmedio', 'Algete',
        'Campo Real', 'Torrejón de la Calzada', 'Colmenar Viejo',
        'Valdemorillo', 'Fuente El Saz de Ja

Conclusiones: Tenemos 168 poblaciones que están listadas en la tupla encima de este markdown.

#### Ejercicio 08

¿Cuál es la media de precios en la población (columna level5) de "Arroyomolinos (Madrid)"? (★★☆)

In [16]:
filtro = df[df['level5'] == 'Arroyomolinos (Madrid)']

media_precios = filtro['price'].mean()

f'La media de precios en Arroyomolinos (Madrid) es de {media_precios:.2f} €'

'La media de precios en Arroyomolinos (Madrid) es de 294541.60 €'

#### Ejercicio 09.

¿Los precios promedios de "Valdemorillo" y "Galapagar" son iguales? (★★☆)

- Muestra ambos promedios
- Escribe en una celda markdown una conclusión sobre ellos

In [21]:
valdemorillo_prices = df[df['level5'] == 'Valdemorillo']['price']

#Calculo el promedio
promedio_valdemorillo = valdemorillo_prices.mean()
f"Promedio Valdemorillo: {promedio_valdemorillo}"


'Promedio Valdemorillo: 363860.2931034483'

In [22]:
galapagar_prices = df[df['level5'] == 'Galapagar']['price']
promedio_galapagar = galapagar_prices.mean()
f"Promedio Galapagar: {promedio_galapagar}",

('Promedio Galapagar: 360063.20238095237',)

#### Ejercicio 10

¿Los promedios de precio por metro cuadrado (precio/m2) de "Valdemorillo" y "Galapagar" son iguales? (★★☆)

> Pista: Crea una nueva columna llamada `pps` (*price per square* o precio por metro cuadrado) y luego analiza los valores.

- Muestra ambos promedios de precio por metro cuadrado
- Escribe en una celda markdown una conclusión sobre ellos

In [25]:
df['pps'] = df['price'] / df['surface']

valdemorillo_pps = df[df['level5'] == 'Valdemorillo']['pps']
promedio_valdemorillo_pps = valdemorillo_pps.mean()
f"Promedio pps Valdemorillo: {promedio_valdemorillo_pps}"

'Promedio pps Valdemorillo: 1317.9502109024986'

In [26]:
df['pps'] = df['price'] / df['surface']

galapagar_pps = df[df['level5'] == 'Galapagar']['pps']
promedio_galapagar_pps = galapagar_pps.mean()
f"Promedio pps Galapagar: {promedio_galapagar_pps}"

'Promedio pps Galapagar: 1606.3240303094024'

#### Ejercicio 11

¿Cuántas agencia de bienes raíces contiene el dataset? (★★☆)

- Muestra el valor obtenido.

In [27]:
num_agencias = df['realEstate_name'].nunique()

f"Número de agencias de bienes raíces: {num_agencias}"

'Número de agencias de bienes raíces: 1821'

#### Ejercicio 12

¿Cuál es la población (columna level5) que contiene la mayor cantidad de casas?(★★☆)

- Muestra la población y el número de casas.

In [29]:
casas_por_poblacion = df['level5'].value_counts()
num_casas_max = casas_por_poblacion.max()
f"Número de casas: {num_casas_max}"


'Número de casas: 6643'

In [30]:
poblacion_max = casas_por_poblacion.idxmax()
f"Población con más casas: {poblacion_max}"

'Población con más casas:  Madrid Capital'

---

## Trabajando con un subconjunto del DataFrame

#### Ejercicio 13

Ahora vamos a trabajar con el "cinturón sur" de Madrid.

Haz un subconjunto del DataFrame original que contenga las siguientes poblaciones (columna level5): "Fuenlabrada", "Leganés", "Getafe", "Alcorcón" (★★☆)

> Pista: Filtra el DataFrame original usando la columna `level5` y la función `isin`.

In [31]:
cinturon_sur = ["Fuenlabrada", "Leganés", "Getafe", "Alcorcón"]

df_cinturon_sur = df[df['level5'].isin(cinturon_sur)]
df_cinturon_sur.head()

Unnamed: 0  id_realEstates  isNew                    realEstate_name  \
1             2       153867863  False  tecnocasa fuenlabrada ferrocarril   
3             4       152776331  False  tecnocasa fuenlabrada ferrocarril   
85           86       153152077  False            sinergical inmobiliaria   
94           95       153995577  False                    viviendas365com   
109         110       153586414  False    area uno asesores inmobiliarios   

     phone_realEstate                                       url_inmueble  \
1         916358736.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
3         916358736.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   
85                NaN  https://www.fotocasa.es/es/comprar/vivienda/le...   
94        911226014.0  https://www.fotocasa.es/es/comprar/vivienda/le...   
109       912664081.0  https://www.fotocasa.es/es/comprar/vivienda/ma...   

     rooms  bathrooms  surface   price  ... level5Id level6Id level7Id  \
1      3.0        1.0      NaN   89000  ...        0        0        0   
3      3.0        1.0     86.0   89000  ...        0        0        0   
85     1.0        1.0     50.0  107000  ...        0        0        0   
94     3.0        2.0    120.0  320000  ...        0        0        0   
109    3.0        3.0    142.0  425000  ...        0        0        0   

    level8Id accuracy          latitude          longitude zipCode customZone  \
1          0        1          40,28674           -3,79351     NaN        NaN   
3          0        0  40,2853785786438  -3,79508142135624     NaN        NaN   
85         0        1          40,35059           -3,82693     NaN        NaN   
94         0        0          40,31933           -3,77574     NaN        NaN   
109        0        0        40,3313411         -3,8313868     NaN        NaN   

             pps  
1            NaN  
3    1034.883721  
85   2140.000000  
94   2666.666667  
109  2992.957746  

[5 rows x 38 columns]

#### Ejercicio 14

Calcula la media y la varianza de muestra para las siguientes variables: precio, habitaciones, superficie y baños (★★★)

> Debes usar el subset obtenido en la pregunta 13

- Crea y visualiza un diccionario con todos los valores

In [32]:
columnas = ['price', 'rooms', 'surface', 'bathrooms']

estadisticas = {}

for col in columnas:
    media = df_cinturon_sur[col].mean()
    varianza = df_cinturon_sur[col].var(ddof=1)
    estadisticas[col] = {'media': media, 'varianza': varianza}

estadisticas

{'price': {'media': np.float64(223094.48070562293),
  'varianza': np.float64(14921367508.04902)},
 'rooms': {'media': np.float64(3.0177383592017737),
  'varianza': np.float64(0.7188858892927542)},
 'surface': {'media': np.float64(111.75222363405337),
  'varianza': np.float64(4263.051760316337)},
 'bathrooms': {'media': np.float64(1.633221850613155),
  'varianza': np.float64(0.5717968625577321)}}

#### Ejercicio 15

¿Cuál es la casa más cara de cada población del cinturón sur de Madríd? (★★☆)

> Debes usar el subset obtenido en la pregunta 13

- Genera un DataFrame con esta información
- Muestra tanto la dirección como el precio de la casa seleccionada de cada población.
- Genera conclusiones en una celda markdown

In [33]:
idx_max_precio = df_cinturon_sur.groupby('level5')['price'].idxmax()
casas_mas_caras = df_cinturon_sur.loc[idx_max_precio, ['level5', 'address', 'price']]
casas_mas_caras = casas_mas_caras.reset_index(drop=True)

casas_mas_caras

level5                                address    price
0     Alcorcón                               Alcorcón   950000
1  Fuenlabrada  Calle de Paulo Freire, 5, Fuenlabrada   490000
2       Getafe                                 Getafe  1050000
3      Leganés           Avenida Reina Sofía, Leganés   650000

Mi conclusión:
Cada población del cinturón sur de Madrid tiene casas con precios máximos distintos.
Estas propiedades representan los extremos del mercado en cada área:
- Fuenlabrada: Calle de Paulo Freire, 5, Fuenlabrada, precio 490000 €
- Leganés: Avenida Reina Sofía, Leganés, precio 650000 €
- Getafe: Getafe, precio 1050000 €
- Alcorcón: Alcorcón, precio 950000 €

Vemos que el precio de la casa más cara varía significativamente entre poblaciones, reflejando diferencias de demanda y características del mercado local.

#### Ejercicio 16

¿Qué puedes decir acerca del precio por metro cuadrado (precio/m2) entre los municipios de 'Getafe' y 'Alcorcón'? (★★☆)

> Debes usar el subset obtenido en la pregunta 13
>
> Pista: Crea una nueva columna llamada `pps` (price per square en inglés) y luego analiza los valores

In [34]:
pps_getafe = round(df_cinturon_sur[df_cinturon_sur['level5'] == 'Getafe']['pps'].mean(), 2)
f"Precio promedio por m² en Getafe: {pps_getafe}"

'Precio promedio por m² en Getafe: 2066.31'

In [35]:
pps_alcorcon = round(df_cinturon_sur[df_cinturon_sur['level5'] == 'Alcorcón']['pps'].mean(), 2)
f"Precio promedio por m² en Alcorcón: {pps_alcorcon}"

'Precio promedio por m² en Alcorcón: 2239.3'

Comparando el precio por metro cuadrado (pps) entre Getafe y Alcorcón:
- Getafe: 2066,31 €/m²
- Alcorcón: 2239.3€/m²

Veo que el precio por m² es mayor en Alcorcón.
Esto indica diferencias en la valorización del suelo y la demanda inmobiliaria entre ambos municipios.


## Conclusiones

#### Ejercicio 17

El análisis muestra que el mercado inmobiliario no es homogéneo.
Los precios, el tamaño de las viviendas y la valorización por metro cuadrado dependen en gran escala de la población y la ubicación. Esto resalta la importancia de segmentar el análisis por municipio y por características de las propiedades para obtener conclusiones precisas y útiles.